src/models/Visitors.php

```diff
@@ -0,0 +1,344 @@
+    {
+        $visitorSession = new VisitorSession();
+        $visitorSession->visitor_id = $visitorId;
+        $visitorSession->session_id = $this->id;
+        $visitorSession->started_at = $this->startedAt();
+        $visitorSession->last_action_at = $this->lastActionAt();
+        $visitorSession->ip = $this->getIpUser();
+        $visitorSession->first_visited_page_id = $this->getPageId(Yii::$app->request->url);
+        $visitorSession->first_activity_at = $this->firstActivityAt();
+        $visitorSession->last_visited_page_id = $this->getPageId(Yii::$app->request->url);
+        $visitorSession->last_activity_at = $this->lastActivityAt();
+
+//        TODO add data param in the next iteration or for the future
+//        $visitorSession->intents_count;
+//        $visitorSession->actions_count;
+//        $visitorSession->goals_count;
+//        $visitorSession->actions_value;
+//        $visitorSession->goals_value;
+//        todo When traffic uncommented result
+        $visitorSession->traffic_sources_id = 0;
+        $visitorSession->save();
+
+        Visitor::updateAll([
+            'sessions_count' => count(VisitorSession::findAll(['visitor_id' => $visitorId]))
+        ], [
+            '=', 'id', $visitorId
+        ]);
+
+        return $visitorSession->id;
+    }
+
+    /**
+     * @return mixed|string
+     */
+    private function getReferrer()
+    {
+        return Yii::$app->request->referrer;
+    }
+
+    /**
+     * @return array
+     */
+    public function init()
+    {
+        if (!$this->get($this::VISITOR_NAME_KEY, false) && !$this->get($this::SESSION_NAME_KEY, false)) {
+            if (!$this->getCookieValue($this::KEY_COOKIE)) {
+                $visitorId = $this->addCookieReturnValue($this::KEY_COOKIE, $this->addVisitor(), $this->timeExpire());
+                $sessionId = $this->addSession($visitorId);
+            } else {
+                $visitorId = Visitor::findOne(['id' => $this->getCookieValue($this::KEY_COOKIE)])->id;
+                $sessionId = VisitorSession::findOne([
+                    'visitor_id' => $visitorId,
+                    'session_id' => $this->id
+                ]);
+
+                if ($sessionId === null) {
+                    $sessionId = $this->addSession($visitorId);
+                }
+            }
+
+            $this->set($this::VISITOR_NAME_KEY, $visitorId);
+            $this->set($this::SESSION_NAME_KEY, $sessionId);
+
+            return [$visitorId, $sessionId];
+        } else {
+            return [
+                $this->get($this::VISITOR_NAME_KEY, false),
+                $this->get($this::SESSION_NAME_KEY, false)
+            ];
+        }
+    }
+
+    /**
+     * @param $link
+     * @return int|null|string
+     */
+    public function getPageIdFromDB($link)
+    {
+        $page = new VisitedPage();
+        $checkPage = $page->findOne(['url' => $link]);
+
+        if ($link !== null && empty($checkPage)) {
```

Comment:
anyway, you only need to create record for URL or get the id of already existing record